snowfall/training/ctc_graph.py

```diff
@@ -1,39 +1,64 @@
 # Copyright (c)  2020  Xiaomi Corp.       (author: Fangjun Kuang)
 
 from functools import lru_cache
-from typing import (
-    Iterable,
-    List,
-)
+from typing import Iterable
+from typing import List
 
 import torch
 import k2
 
 
+def create_bigram_phone_lm(phones: List[int]) -> k2.Fsa:
+    '''Create a bigram phone LM.
+    The resulting FSA (P) has a start-state and a state for
+    each phone 0, 1, ....; and each of the above-mentioned states
+    has a transition to the state for each phone and also to the final-state.
+
+    Caution:
+      blank is not a phone.
+    '''
+    final_state = len(phones) + 1
+    rules = ''
+    for i in range(1, final_state):
+        rules += f'0 {i} {phones[i-1]} 0.0\n'
```

Comment:
I’m wondering if it’s not more convenient in the long run to build FSAs/FSTs programmatically with an OpenFST-like interface: Fsa.add_state(), Fsa.add_arc()

Reply:
mm, fundamentally k2 FSAs aren't mutable so we'd have to have some alternative data structure underlying that.  Possible though.